### PR TITLE
perf: kmean fit uses cached torch dataset

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -418,7 +418,7 @@ def sample(
         table : Table
         """
         total_num_rows = self.count_rows()
-        indices = random.sample(range(total_num_rows), num_rows)
+        indices = sorted(random.sample(range(total_num_rows), num_rows))
```
Comment (QianZhu) [inline python/python/lance/dataset.py:421]: add explanation on the benefits of indices being sorted

Comment (wjones127) [reply python/python/lance/dataset.py:421]: This is a breaking API change. Could we add an argument `sorted: bool=False` to this function so users can optionally enable this?

File: python/python/lance/torch/data.py [hunk 2]
```diff
@@ -76,14 +85,19 @@ def __init__(
...
             raise ValueError("Does not support sampling over filtered dataset")
 
         self.cache = cache
         self.cached_ds: Optional[CachedDataset] = None
 
+    def __repr__(self) -> str:
+        return f"LanceTorchDataset({self.dataset.uri}, size={self.samples})"
+
     def __iter__(self):
+        print("LanceDataset::__iter__: samples=", self.samples)
```
Comment (QianZhu) [inline python/python/lance/torch/data.py:100]: delete?

File: python/python/lance/dataset.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/dataset.py:421]: `sorted` indicates to sort by value. Shall we use `sequential` or sth ?

Comment (wjones127) [reply python/python/lance/dataset.py:421]: Perhaps `randomize_order: bool=True`? Then if `randomize_order` is false we call `sorted`. What do you think of that?

Comment (eddyxu) [reply python/python/lance/dataset.py:421]: Ok, it sounds better.

**Comment by eddyxu** [comment]
Trained on IVF65535,PQ192 on a Titan RTX,  finished IVF part training in `2390`s.

**Review by westonpace — APPROVED** [review]
Seems like python's gc isn't working very well.  I wonder if it has any way of calculating memory pressure for native objects.

File: python/python/lance/sampler.py [hunk 3]
```diff
@@ -63,33 +67,35 @@ def _efficient_sample(
...
+                list(range(offset, offset + chunk_sample_size)),
                 columns=columns,
-                batch_size=chunk_size,
-                offset=offset,
-                limit=chunk_sample_size,
-            )
+            ).to_batches()
         )
+        if idx % 50 == 0:
+            logging.info("Sampled at offset=%s, len=%s", offset, chunk_sample_size)
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/python/lance/sampler.py:80]: Yea, this wanted to show progress during long training.

**Pull request merged**